Chore: Guides (#25)
Interactions-as-a-Service/d1-orm: guides/models.md

Models are a wrapper around the [Query Builder](/guides/query-building) that allows you to manage the queries and data in a more object-oriented way.

### Creating a Model

To start with, we'll make a type for our model. This is optional, but it's recommended to make a type for your model so you can have improved type safety.

```ts
type User = {
	id: number;
	name: string;
	email: string | undefined;
};
```

Now we need to define the Model structure.

```ts
import { D1Orm, DataTypes, Model } from "d1-orm";

> // We must initialise an ORM to use the Model class. This is done by passing in a D1Database instance.

Reviewer: ```suggestion
// We must initialise an ORM to use the Model class. This is done by passing in a D1Database instance (in this case it's bound to the `env.DB` environment variable).
```

const orm = new D1Orm(env.DB);

// Now, to create the model:
const users = new Model<User>(
	{
		D1Orm: orm,
		tableName: "users",
	},
	{
		id: {
			type: DataTypes.INTEGER,
			primaryKey: true,
			autoIncrement: true,
			notNull: true,
		},
		name: {
			type: DataTypes.STRING,
			notNull: true,
			defaultValue: "John Doe",
		},
		email: {
			type: DataTypes.STRING,
			unique: true,
		},
	}
);
```

Both arguments are required. The first argument is the model options, and the second argument is the model schema.

- The model options should contain the `D1Orm` instance we created earlier, and the `tableName` of the model.

- The model schema should contain the columns of the model. The key of the object should be the name of the column, and the value should be the column options. See {@link ModelColumn} for API reference.

The next step is to create the table in the database. This is done by calling the {@link Model.CreateTable} method on the model. There are two strategies for this:

- default: This will attempt to create the table, and error if it already exists.

> - force: This will drop the table if it already exists, and then recreate it. Warning: This will delete all data in the table.

Reviewer: ```suggestion
- force: This will drop and recreate the table. Warning: This will delete all data in the table!
```
Made warning more forceful and removed "if it exists" because it's implied.

To use it, it's as simple as

```ts

> await users.CreateTable({ strategy: "default" /* or "force" */ });

Reviewer: ```suggestion
await users.CreateTable({ strategy: "default" /* or "force", see above */ });
```

```

This will either return successfully or throw an error that your table already exists.

It's not recommended to use the force strategy in production, but it's useful for development.

> You additionally shouldn't call this method on each worker request, as it's an expensive operation. Instead, you should call it once when you deploy your worker.

Reviewer: ```suggestion
You shouldn't call this method on each worker request: it's very expensive and may take some time. Instead, it should be called when you deploy your Worker.

Alternatively, you can create your tables manually via the [`wrangler` CLI](https://developers.cloudflare.com/workers/wrangler/) and avoid using this method all together.
```

Reviewer: Added alternative, as some people may prefer to create tables via their own SQL via Wrangler instead of via a JS command at deploy time (may be tricky to access a binding at deploy time).
